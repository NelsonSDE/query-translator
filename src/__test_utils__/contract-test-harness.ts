import { INestApplication, ValidationPipe } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { TestingModule } from '@nestjs/testing';
import { AppModule } from '../app.module';
import * as request from 'supertest';
import { HttpExceptionFilter } from '../common/filters/http-exception.filter';
export class ContractTestHarness {
  app: INestApplication;

  async sendPostAndExpectSuccess(
    endpoint: string,
    payload: any,
    statusCode = 200,
  ) {
    const response = await request(this.app.getHttpServer())
      .post(endpoint)
      .send(payload);

    expect(response.status).toBe(statusCode);
    expect(response.body).toMatchSnapshot();
    return response.body;
  }

  async sendPostAndExpectError(
    endpoint: string,
    payload: any,
    statusCode = 400,
  ) {
    const response = await request(this.app.getHttpServer())
      .post(endpoint)
      .send(payload);

    expect(response.status).toBe(statusCode);
    expect(response.body).toMatchSnapshot();
    return response.body;
  }
}

const testHarness = new ContractTestHarness();

export function withContractTestHarness() {
  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    const app = moduleFixture.createNestApplication();
    app.useGlobalPipes(
      new ValidationPipe({
        transform: true,
        whitelist: true,
        forbidNonWhitelisted: true,
        validationError: { target: true, value: true },
        validateCustomDecorators: true,
        forbidUnknownValues: false,
      }),
    );
    app.useGlobalFilters(new HttpExceptionFilter());
    await app.init();
    testHarness.app = app;
  });

  afterAll(async () => {
    await testHarness.app.close();
  });
  return testHarness;
}
