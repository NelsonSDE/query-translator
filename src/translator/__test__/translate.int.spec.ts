import { Language } from '../enums/language.enum';
import { withContractTestHarness } from '../../__test_utils__/contract-test-harness';

const testHarness = withContractTestHarness();

describe('TranslateController (e2e)', () => {
  describe('POST /translate', () => {
    it('should translate MongoDB query to SQL', async () => {
      const mongoQuery = {
        query: 'db.users.find({age: {$gt: 18}});',
        source: Language.MONGO_DB,
        target: Language.SQL,
      };

      await testHarness.sendPostAndExpectSuccess('/translate', mongoQuery);
    });

    it('should thow a 422 when translating SQL query to MongoDB', async () => {
      const sqlQuery = {
        query: 'SELECT * FROM users WHERE age > 18;',
        source: Language.SQL,
        target: Language.MONGO_DB,
      };

      await testHarness.sendPostAndExpectError('/translate', sqlQuery, 422);
    });

    it('should return 400 for invalid source language', async () => {
      const invalidQuery = {
        query: 'SELECT * FROM users',
        source: 'invalid_language',
        target: Language.MONGO_DB,
      };

      await testHarness.sendPostAndExpectError('/translate', invalidQuery);
    });

    it('should return 400 for invalid target language', async () => {
      const invalidQuery = {
        query: 'SELECT * FROM users',
        source: Language.SQL,
        target: 'invalid_language',
      };

      await testHarness.sendPostAndExpectError('/translate', invalidQuery);
    });

    it('should return 400 for empty query', async () => {
      const invalidQuery = {
        query: '',
        source: Language.SQL,
        target: Language.MONGO_DB,
      };

      await testHarness.sendPostAndExpectError('/translate', invalidQuery);
    });
  });
});
