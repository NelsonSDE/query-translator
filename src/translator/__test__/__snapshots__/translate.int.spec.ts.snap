// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TranslateController (e2e) POST /translate should return 400 for empty query 1`] = `
{
  "error": "Bad Request",
  "message": [
    "query should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`TranslateController (e2e) POST /translate should return 400 for invalid source language 1`] = `
{
  "error": "Bad Request",
  "message": [
    "source must be one of the following values: mongodb, sql",
  ],
  "statusCode": 400,
}
`;

exports[`TranslateController (e2e) POST /translate should return 400 for invalid target language 1`] = `
{
  "error": "Bad Request",
  "message": [
    "target must be one of the following values: mongodb, sql",
  ],
  "statusCode": 400,
}
`;

exports[`TranslateController (e2e) POST /translate should thow a 422 when translating SQL query to MongoDB 1`] = `
{
  "message": "Not supported operation: SQL to MongoDB",
  "statusCode": 422,
}
`;

exports[`TranslateController (e2e) POST /translate should translate MongoDB query to SQL 1`] = `
{
  "originalQuery": "db.users.find({age: {$gt: 18}});",
  "source": "mongodb",
  "target": "sql",
  "translatedQuery": "SELECT * FROM users WHERE $gt > 'age';",
}
`;
